KAFKA-8229: Only change nextCommit if and only if now >= to it

sdreynolds opened the pull request:
Summary
When using the commitRequest api within sink context, the next commit
timer should not advance. Timer should only advance when it is the
cause of the commit. Prior to this change, the time would advance
_each_ time a commit happens -- including when a commit happens
because it was requested by the `Task`. When a `Task` requests a
commit several times, the clock advances far into the future
preventing commits from happening.

This commit changes the behavior, if the `Task` requested a commit and
the `nextCommit` is in the _near_ future, `nextCommit` will not
advance.

KAFKA-8229

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

sdreynolds committed:
Only change nextCommit if and only if now >= to it

Summary
When using the commitRequest api within sink context, the next commit
timer should not advance. Timer should only advance when it is the
cause of the commit. Prior to this change, the time would advance
_each_ time a commit happens -- including when a commit happens
because it was requested by the `Task`. When a `Task` requests a
commit several times, the clock advances far into the future
preventing commits from happening.

This commit changes the behavior, if the `Task` requested a commit and
the `nextCommit` is in the _near_ future, `nextCommit` will not
advance.

KAFKA-8229

hachikuji commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -612,6 +614,11 @@ SinkTaskMetricsGroup sinkTaskMetricsGroup() {
     }
 
+    // Visible for testing
+    public long getNextCommit() {
```
Could be default access if we only need it for testing.

hachikuji commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:
```diff
@@ -206,7 +206,9 @@ protected void iteration() {
             if (!committing && (context.isCommitRequested() || now >= nextCommit)) {
                 commitOffsets(now, false);
-                nextCommit += offsetCommitIntervalMs;
+                if (now >= nextCommit) {
```
Maybe this is a little simpler?
```java
nextCommit = now + offsetCommitIntervalMs;
```

sdreynolds replied:
That changes the system in a different way. So in your suggestion, `commitRequest` will always move the `nextCommit` forward -- but notably not into the _far_ future. 

Vs my original change which only moves `nextCommit` forward if it is in the "past".

Unclear to me after re-reading the [original Kip](https://cwiki.apache.org/confluence/display/KAFKA/KIP-89%3A+Allow+sink+connectors+to+decouple+flush+and+offset+commit) what the intention is for this behavior. 

Both solve the problem we faced. Which behavior do you think is better? Always move `nextCommit` forward on each commit or only move `nextCommit` if it is in the past?

hachikuji replied:
I don't think it makes a big difference either way. The intent of the offset commit interval is just to make sure committed offsets don't fall too far behind. It does not need to be a strict schedule. It seemed more intuitive and simpler to me to reset the interval. In any case, we should get rid of this relative tracking of the next commit. If we use an absolute time, then we will not have problems like this in the future.

hachikuji replied:
@sdreynolds I would suggest we use my suggestion since it addresses the problem with the relative time adjustments. It could also happen that `now` somehow gets far ahead of `nextCommit`. In this case, we we do not want a flurry of useless offset commits in order for `nextCommit` to catch back up. It seems simpler to reset the offset commit interval.

hachikuji replied:
Is there any reason not to accept this suggestion? I will go ahead and push an update to this PR next week if there are no further responses so that we can get this fix into the next release.

hachikuji reviewed (commented):
Thanks, good find. Left a couple small comments.

sdreynolds committed:
remove `public` from `getNextCommit()` method

Summary:
Upon review, it was pointed out it doesn't need to be `public` for the
test to leverage it.

sdreynolds commented:
I agree with your suggestion, I haven't had time to set down and write it.

On Tue, May 21, 2019, 4:00 PM Jason Gustafson <notifications@github.com>
wrote:

> *@hachikuji* commented on this pull request.
> ------------------------------
>
> In
> connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
> <https://github.com/apache/kafka/pull/6579#discussion_r286256599>:
>
> > @@ -206,7 +206,9 @@ protected void iteration() {
>              // Maybe commit
>              if (!committing && (context.isCommitRequested() || now >= nextCommit)) {
>                  commitOffsets(now, false);
> -                nextCommit += offsetCommitIntervalMs;
> +                if (now >= nextCommit) {
>
> Is there any reason not to accept this suggestion? I will go ahead and
> push an update to this PR next week if there are no further responses so
> that we can get this fix into the next release.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/kafka/pull/6579?email_source=notifications&email_token=AIVVLICJ7WHOGIRM4ZYS4CDPWR5ILA5CNFSM4HFZJKJKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZJU2RQ#discussion_r286256599>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIVVLIGGBVBYOR4RDV7NE3TPWR5ILANCNFSM4HFZJKJA>
> .
>

hachikuji commented:
@sdreynolds Thanks. Any time in the next week would be great. Let me know if you need help.

sdreynolds committed:
Move from relative tracking of the next commit to absolute

Summary:
Prior to this change nextCommit would advance without referencing the
current time. This can lead to subtle bugs in the future and by
changing this to use `now`, we are able to ensure we don't advance to
far into the future after several sequential calls to
`context.requestCommit()`.

hachikuji reviewed (approved):
Thanks @sdreynolds. LGTM.
